datalad/auto.py

```diff
@@ -0,0 +1,184 @@
+# emacs: -*- mode: python; py-indent-offset: 4; tab-width: 4; indent-tabs-mode: nil -*-
+# ex: set sts=4 ts=4 sw=4 noet:
+# ## ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ##
+#
+#   See COPYING file distributed along with the datalad package for the
+#   copyright and license terms.
+#
+# ## ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ##
+"""Proxy basic file operations (such as open) to obtain files automagically upon I/O
+"""
+
+from mock import patch
+from six import PY2
+import six.moves.builtins as __builtin__
+builtins_name = '__builtin__' if PY2 else 'builtins'
+
+try:
+    import h5py
+except ImportError:
+    h5py = None
+
+import logging
+
+from os.path import dirname, abspath, pardir, join as opj, exists, basename
+
+from .support.annexrepo import AnnexRepo
+from .support.gitrepo import GitRepo
+from .support.exceptions import CommandError
+from .cmd import Runner
+
+from .utils import swallow_outputs
+lgr = logging.getLogger("datalad.auto")
+
+
+# TODO: shouldn't be some classmethod of GitRepo?
```

Comment:
Well, ls-files has to somehow find git's index too, hasn't it? So, finding the top is a task, that ls-files has to accomplish as well, I guess. Anyway, as long as we don't find a faster search on the index of a GitPython.Repo, let's stay with ls-files.